test/distributed/org/apache/cassandra/distributed/test/sai/IndexStreamingFailureTest.java

```diff
@@ -0,0 +1,160 @@
+    {
+        try (Cluster cluster = init(Cluster.build(2)
+                                           .withConfig(c -> c.with(NETWORK, GOSSIP).set("stream_entire_sstables", streamEntireSSTables))
+                                           .withInstanceInitializer(injectedError)
+                                           .start()))
+        {
+            cluster.disableAutoCompaction(KEYSPACE);
+            cluster.schemaChange(withKeyspace("CREATE TABLE %s.test (pk int PRIMARY KEY, v text)"));
+            cluster.schemaChange(withKeyspace("CREATE CUSTOM INDEX test_v_index ON %s.test(v) USING 'StorageAttachedIndex'"));
+
+            IInvokableInstance first = cluster.get(1);
+            IInvokableInstance second = cluster.get(2);
+
+            first.executeInternal(withKeyspace("INSERT INTO %s.test(pk, v) VALUES (?, ?)"), 1, "v1");
+            first.flush(KEYSPACE);
+
+            Object[][] rs = second.executeInternal(withKeyspace("select pk from %s.test where v = ?"), "v1");
+            assertThat(rs.length).isEqualTo(0);
+
+            // The repair job should fail when index completion fails. This should also fail the streaming transaction.
+            NodeToolResult result = second.nodetoolResult("repair", KEYSPACE);
+            result.asserts().failure().errorContains(TEST_ERROR_MESSAGE);
+
+            // The SSTable should not be added to the table view, as the streaming transaction failed...
+            rs = second.executeInternal(withKeyspace("select pk from %s.test where pk = ?"), 1);
+            assertThat(rs.length).isEqualTo(0);
+
+            // ...and querying the index also returns nothing, as the index for the streamed SSTable was never built.
+            rs = second.executeInternal(withKeyspace("select pk from %s.test where v = ?"), "v1");
+            assertThat(rs.length).isEqualTo(0);
+
+            // On restart, ensure that the index remains querable and does not include the data we attempted to stream. 
+            second.shutdown().get();
+            second.startup();
+
+            // On restart, the base table should be unchanged...
+            rs = second.executeInternal(withKeyspace("select pk from %s.test where pk = ?"), 1);
+            assertThat(rs.length).isEqualTo(0);
+
+            // ...and the index should remain queryable, because from its perspective, the streaming never happened.
+            Boolean isQueryable = second.callOnInstance(() -> {
+                SecondaryIndexManager sim = Keyspace.open(KEYSPACE).getColumnFamilyStore("test").indexManager;
+                Index index = sim.getIndexByName("test_v_index");
+                return sim.isIndexQueryable(index);
+            });
+
+            assertTrue("The index should be queryable on restart!", isQueryable);
+            rs = second.executeInternal(withKeyspace("select pk from %s.test where v = ?"), "v1");
+            assertThat(rs.length).isEqualTo(0);
```

Comment:
Maybe we could extend the test case a bit by deactivating the byteman rule and running a second repair. Then we would verify that this time the data arrives at the second node and it's queryable. Something like:
```java
second.runOnInstance(()-> ByteBuddyHelper.fail.set(false));
result = second.nodetoolResult("repair", KEYSPACE);
result.asserts().success();

rs = second.executeInternal(withKeyspace("select pk from %s.test where v = ?"), "v1");
assertThat(rs.length).isEqualTo(1);
```
This would be just to ensure we don't enter into some kind of irreversible state after a failure, if that makes sense.